[FLINK-13938][yarn] Use pre-uploaded flink binary to accelerate flink submission

wangyang0918 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
Currently, every time we start a flink cluster, flink lib jars need to be uploaded to hdfs and then register Yarn local resource so that it could be downloaded to jobmanager and all taskmanager container. I think we could have two optimizations.
1. Use pre-uploaded flink binary to avoid uploading of flink system jars
2. Use pre-uploaded flink binary as yarn public cache to eliminate the unnecessary jars downloading. The public cache could be shared by different applications.

How to use the shared lib feature?
1. Upload a copy of flink release binary to hdfs.
2. Use the -ypl argument to specify the shared lib

`./bin/flink run -d -m yarn-cluster -p 20 -ypl hdfs://hdpdev/flink/release/flink-1.9.0 examples/streaming/WindowJoin.jar`

https://issues.apache.org/jira/browse/FLINK-13938

## Brief change log

* Upload a copy of Flink binary beforehand and specify the path to avoid uploading local resources.
* Register pre-uploaded resources as yarn public cache


## Verifying this change

* UT test, a lib directory existence should be check, `YARNITCase#testPreUploadedFlinkPath`
* Manually check, start a flink cluster with `-ypl`, it should be very fast.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -149,61 +155,79 @@ public static void setupYarnClassPath(Configuration conf, Map<String, String> ap
-				localSrcPath);
-		}
+		String relativeTargetPath,
+		@Nullable FileStatus preUploadedFileStatus) throws IOException {
```
I think we should separate register pre-uploaded files into another method instead just hacking a huge if-else block. If you take a close look you will find this two code path share little code and pre-uploaded path even doesn't use most of the parameter.

wangyang0918 replied:
I have moved the `setupLocalResource` without preUploadedFileStatus to another method to  avoid huge if-else block.

TisonKun replied:
It seems that you don't push the commit so far. One reminder is for smooth review you can push the change in another commit :-)

walterddr replied:
+1, how about splitting it into: `setUpLocalResource()` and `setUpPreloadedResource` --> since `Local` really referring to something that's "local" to the submission client that runs this `Utils.class`

BTW I really like how @wangyang0918 split the setup and the register part out! e.g. `registerLocalResource` (renaming it to `registerResource` might've been even better)

wangyang0918 replied:
Thanks @TisonKun and @walterddr. It makes sense to me. I will split in into two methods `setupLocalResource()` and `setupPreUploadedResource`.

However, i suggest not to change the name `registerLocalResource`. Since the local not means the local file, it means the Yarn `LocalResource`. It is a proper noun of Yarn.

```
LocalResource localResource = Records.newRecord(LocalResource.class);
```

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -509,7 +536,8 @@ static ContainerLaunchContext createTaskExecutorContext(
 					homeDirPath,
-					"").f1;
+					"",
+					null).f1;
```
With my comment above we don't need to pass most of `null` default as here and else where.

wangyang0918 replied:
I think `null` means we always do not need to use pre-uploaded file for this local file. If you prefer add a new function and check not null outside of `setupSingleLocalResource`, i will do as you say.

walterddr replied:
I think what @TisonKun means is if we use `setUpLocalResource()` and `registerPreloadedResource` as 2 different function. it is not needed to even pass in the null checker.

wangyang0918 commented:
@TisonKun I just want to redeclare that this PR will have two optimizations.
1. Use pre-uploaded flink jars to avoid uploading in flink client, it will make the submission very fast. So the corresponding directory will be needed.
2. Register pre-uploaded flink jars as Yarn public cache. This will make the public cache could be shared by different applications and avoid unnecessary downloading.

I suggest to upload a flink binary to hdfs first, and then the two optimizations will both work as expected. 

If you want to support multiple paths, i just want to know how it works? How to find the corresponding local file and avoid uploading?

TisonKun commented:
Briefly speaking we are in different context.

- What you want to do is that files are accessible locally but prevent to upload if they are already pre-uploaded.
- What I want to do is that files are possibly no exist locally at all so that I add resources remotely.

You changes implement two step (1) include remote resources (2) exclude local resources in one opinion. I agree we can do it by filtering by name but it is beyond my requirement.

Again, your concern is more like performance but my concern is about permission(availability).

TisonKun commented:
For multiple paths, you just do as local ship file does; i.e., iterate every path/file and set local resources respectively. I don't consider much on exclusion things here.

wangyang0918 commented:
After discussion with @TisonKun offline, we realize that [PR-10187](https://github.com/apache/flink/pull/10187) is another issue, he will create a new JIRA to track. It is `-yt` could support a remote distributed directory and will not need public distributed cache. For example, '-yt hdfs://nn1/path1'.

This PR is focused on using pre-uploaded flink binary to accelerate submission and container launch.

TisonKun commented:
I don't think `-yt hdfs://` and specially handle it is a good way but let's discuss this in another issue.

walterddr commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -263,6 +298,36 @@ private static LocalResource registerLocalResource(FileSystem fs, Path remoteRsr
+	 * @param yarnConfig yarn configuration
+	 * @return local resource tuple, f0 is filename, f1 is local resource.
+	 */
+	private static Tuple2<String, LocalResource> registerLocalResource(
```
can also split into `registerLocalResource` and `registerPreloadedResource` - or if we think this is not reused somewhere else, we can always merge it with the `setUp...` functions

wangyang0918 replied:
I prefer the name `registerLocalResource`. It has different overloadings.

```
private static LocalResource registerLocalResource(Path remoteRsrcPath, long resourceSize, long resourceModificationTime, LocalResourceVisibility visibility)

private static LocalResource registerLocalResource(FileSystem fs, Path remoteRsrcPath)

private static Tuple2<String, LocalResource> registerLocalResource(String resourceDesStr, Configuration yarnConfig)
```

TisonKun replied:
This method is quite tricky for introducing an in-place pattern "RemotePath;[resourceSize;resourceModificationTime;LocalResourceVisibility]" to parse configuration...I think it works as is but +0 for this patch. @walterddr you can merge this patch if you think it is good to go.

walterddr replied:
@TisonKun: sorry could you elaborate the "in-place" pattern you mentioned? and are you a +0 on the current PR implementation of on the split to `registerLocal...` & `registerPreloaded...` ?

walterddr replied:
@TisonKun correct me if I were wrong. I think you are also having concerns regarding the branch out paradigm here

Reiterating my previous comment: since the code path to deal with preuploaded resource is different from the local resource (given the if-else pattern) we can probably split it into different methods? what do you think @TisonKun @wangyang0918

wangyang0918 commented:
@TisonKun @walterddr 
Sorry for late and really thanks for your review. I have addressed all the comments. Please take a look.

I remember some users in the ML also want this feature. So that it could make the flink submission more smoothly. Also we could avoid the unnecessary downloading jars from hdfs and accelerate the taskmanager launch.

In addition, i have tested this new feature in my Yarn cluster. I works just as expected. Could you help to double check?

walterddr commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -129,6 +134,33 @@ public static void setupYarnClassPath(Configuration conf, Map<String, String> ap
+	 *
+	 * @return Path to remote file (usually hdfs)
+	 */
+	static Tuple2<Path, LocalResource> setupPreUploadedResource(
```
@TisonKun I am not sure if there's a follow up JIRA created yet for PR #10187 ? are you planning to use similar approach like this one to support a remote URL resource (like Jars)? if yes I think this is fine, otherwise `setUpYarnPublicCacheResource` might be more accurate?

wangyang0918 replied:
The preUploaded Resources means they have been uploaded to hdfs already, not Yarn public cache.
I have to make clear that Yarn public cache is a type of `LocalResourceVisibility`. The public cache will be downloaded only once in the same NodeManager and could be shared by different applications.

TisonKun replied:
Not yet. I don't have too much cycles for generally thinking a considerate solution of this demand now. Will revisit it later.

wangyang0918 commented:
@walterddr Your assumption 1 is right. For assumption 2, the pre-uploaded flink binary is located on a generic location of any kind of DFS system(usually HDFS), not Yarn distributed cache.

We register the pre-uploaded flink jars as Yarn public cache, so that it could be shared by different applications. And they will not be downloaded every time.

For #10187, @TisonKun will create a separate JIRA. It is focused on support shipping directory located on HDFS.

walterddr commented:
thanks for the explanation @wangyang0918 :-) --Rong

wangyang0918 commented:
@walterddr @TisonKun Could you please help with the final verification and merge. Thanks.

TisonKun commented on flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNITCase.java:
```diff
@@ -93,8 +114,8 @@ private void deployPerjob(YarnConfigOptions.UserJarInclusion userJarInclusion) t
 			yarnClusterDescriptor.setLocalJarPath(new Path(flinkUberjar.getAbsolutePath()));
-			yarnClusterDescriptor.addShipFiles(Arrays.asList(flinkLibFolder.listFiles()));
-			yarnClusterDescriptor.addShipFiles(Arrays.asList(flinkShadedHadoopDir.listFiles()));
+			yarnClusterDescriptor.addShipFiles(Collections.singletonList(flinkLibFolder));
```
I don't stick to revert this but want to know that, we don't change anything by this diff right? IIRC Flink will iterate over a directory.

wangyang0918 replied:
Yes, you are right. Flink will iterate over a directory. So i just make the test to keep the same behavior with `YarnClusterDescriptor`. `YarnClusterDescriptor` add the `lib` folder as ship file, not all the files.

We find the corresponding pre-uploaded file by the same relative path with local file. So it is important to test the real behavior.

TisonKun commented on flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNITCase.java:
```diff
@@ -129,11 +150,20 @@ private void deployPerjob(YarnConfigOptions.UserJarInclusion userJarInclusion) t
+		final FileSystem fs = FileSystem.get(YARN_CONFIGURATION);
+		final Path stagingDirectory = new Path(fs.getHomeDirectory(), ".flink/" + appId.toString());
+		// If pre-uploaded flink is set correctly, the lib directory will not be uploaded to staging directory.
+		assertEquals(shouldExist, fs.exists(new Path(stagingDirectory, flinkLibFolder.getName())));
```
We can move the comment as `message` parameter of `assertEquals`.

wangyang0918 replied:
Nice suggestion. I will move the message to `assertEquals`.

wangyang0918 commented:
@TisonKun I have addressed your comments. Please take a look.

wangyang0918 commented:
@TisonKun Thanks for your suggestion. I will try to find a more elegant way.

wangyang0918 commented:
@walterddr I think @TisonKun means we should not use the magic code to parse the string `if (resourceDesc.length == 4) ... else ...` in `Utils`. I will add new class `YarnLocalResourceDescription` to describe the data structure and try to simplify the code.

BTW, i think the feature freeze of 1.10 is Dec 8. And we have not missed it :)

wangyang0918 commented:
@TisonKun I have updated the PR with a `fixup`. The magic and confusing codes have been removed. I add a new class `YarnLocalResourceDescription` to describe the local resource information.

Please take a look.

walterddr commented:
> @walterddr I think @TisonKun means we should not use the magic code to parse the string `if (resourceDesc.length == 4) ... else ...` in `Utils`. I will add new class `YarnLocalResourceDescription` to describe the data structure and try to simplify the code.
> 
> BTW, i think the feature freeze of 1.10 is Dec 8. And we have not missed it :)

Oops... sorry for misread the ML discussion, yes you are right it is Dec.8.

@wangyang0918 thanks for the explanation. now it makes sense to me :-) In this case I think we should definitely target getting this in asap.

walterddr commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -907,19 +900,19 @@ private ApplicationReport startAppMaster(
 
 		String flinkConfigKey = "flink-conf.yaml";
-		Path remotePathConf = setupSingleLocalResource(
+		YarnLocalResourceDescription localResourceDescConf = setupSingleLocalResource(
```
makes naming consistent? maybe: `localResourceDescFlinkConf `?

wangyang0918 replied:
Make sense to me. I will update.

walterddr commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnLocalResourceDescription.java:
```diff
@@ -0,0 +1,99 @@
+ * Yarn local resource description is generated by {@link YarnClusterDescriptor} and set to environment of job manager.
+ * Then it will be used to register local resources for task manager in {@link Utils#createTaskExecutorContext}.
+ */
+class YarnLocalResourceDescription {
```
maybe `YarnLocalResourceDescriptor`? **Descriptor** makes more sense to me

wangyang0918 replied:
Make sense to me. I will update.

walterddr commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnLocalResourceDescriptionTest.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Test
+	public void testFromString() throws Exception {
+		final String key = "flink.jar";
```
could you add a test for malformed string that doesn't matches the regex pattern?

wangyang0918 replied:
Of course. I will add a test for malformed pattern.

wangyang0918 commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -574,12 +559,10 @@ static ContainerLaunchContext createTaskExecutorContext(
 
 		// prepare additional files to be shipped
-		for (String shipResourceDescStr : shipListString.split(",")) {
+		for (String shipResourceDescStr : shipListString.split(";")) {
```
I think the `ENV_CLIENT_SHIP_FILES` is an internal environment. The users could not set the ship files via `system.env()`. The only way to set ship files is using `-yt`. And the separator of `-yt` is same as before.
If you insist, i will change `;` to `,`.

wangyang0918 commented:
@walterddr 
I have add a new fixup to address your comments. Please take a look.

walterddr commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -253,6 +253,13 @@
+		key("yarn.pre-uploaded.flink.path")
+			.stringType()
+			.noDefaultValue()
+			.withDescription("Specify the pre-uploaded path of flink. It will be used to accelerate the job submission" +
```
```
Specifies the pre-uploaded path of the Flink Jar. If specified, Flink will locate the pre-uploaded Flink Jar via the provided path during container localization. This could accelerate the job submission process.
```

wangyang0918 replied:
I will change the description.

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/Utils.java:
```diff
@@ -596,6 +644,30 @@ static ContainerLaunchContext createTaskExecutorContext(
+					}
+				}
+			} catch (IOException e) {
+				LOG.warn("Error to get the ");
```
@walterddr @wangyang0918 please fix this log when merging.

walterddr replied:
oops. missed this one.
```
LOG.warn("Error getting pre-uploaded Flink binary via path: {}", flinkSharedBinary, e);
```

wangyang0918 replied:
Nice catch. I will fix it.

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -155,6 +156,10 @@
 
 	private String zookeeperNamespace;
 
+	// Pre-uploaded flink files, key is relative path(for example, lib/flink-dist_{version}.jar),
```
nit: can be doc comment.

wangyang0918 replied:
Will fix it.

TisonKun commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:
```diff
@@ -155,6 +156,10 @@
 
+	// Pre-uploaded flink files, key is relative path(for example, lib/flink-dist_{version}.jar),
+	// and value is remote file status.
+	private final Map<String, FileStatus> preUploadedFlinkFiles;
```
nit: follow the other configs pattern, i.e., init a empty map and call `addAll` when deploy instead of initializing in constructor. It can be a follow-up we generally refactor deployment inside YarnClusterDescriptor.

wangyang0918 replied:
Will fix it.

wangyang0918 commented:
@TisonKun @walterddr Thanks a lot for your review. I have added a fixup to address all your comments. Please take a look.

walterddr commented:
> I'm ok with this patch. But one more concern
> 
> cc @aljoscha @zentol what is backward compatibility we keep in command-line options? I'd like to mark this feature as experimental because some of our users might be looking forward to this improvement in 1.10 but I have an idea generally handle so-called "remote ship files" configuration.
> 
> The problem is the logic we exclude local lib files is based on name matching which is brittle and actually we can achieve the same behavior by configuring both remote ship files and excluding local lib from being uploaded.

Thanks for bringing the point of "backward compatibility" up and actually I also shared the same concern with you when I bring up the question of #10187 - if there's a generic way of handling "remote ship files" consistently we should at least discuss this. @TisonKun would you mind start a JIRA discussion to share some of your thoughts?

wangyang0918 commented:
@TisonKun @walterddr I agree with you that there's a more generic way of handling remote ship files. However, i think it is beyond t'h

walterddr commented:
hmm. I think I am agreeing with @TisonKun here - if there's a more generic way of handling it. it doesn't seem make sense to have `-ypl` CLI option being there for only one release then replacing it with more generic solution. this actually impacts user experiences. 

However I dont think discussing the generic solution of remote resource is suitable in this PR comment section. How about we open a separated JIRA for discussion?

aljoscha commented:
@TisonKun to answer your question: in general we try to not break command line parameters.

kl0u commented:
Hi @wangyang0918 ! I am closing this PR as it has been subsumed by https://github.com/apache/flink/pull/12040
